lib/rules/prefer-lookaround.ts

```diff
@@ -0,0 +1,625 @@
+type ReplaceReferences = {
+    // Reference at the starting position.
+    // e.g.
+    // '$1str$2' -> { ref: 1, range: [0,2] }
+    // '$<foo>str$<bar>' -> { ref: 'foo', range: [0,6] }
+    // 'str$1' -> null
+    // 'str$1str$2' -> null
+    startRef: ReplaceReference | null
+    // Reference at the ending position.
+    // e.g.
+    // '$1str$2' -> { ref: 2, range: [5,7] }
+    // '$<foo>str$<bar>' -> { ref: 'bar', range: [9,15] }
+    // '$1str' -> null
+    // '$1str$2str' -> null
+    endRef: ReplaceReference | null
+    // All references including the above.
+    allRefs: ReplaceReference[]
+}
+
+/**
+ * Holds all replacement reference data.
+ *
+ * If the same RegExp instance is used for replacement in 2 places, the number of data in `list` is 2.
+ */
+class ReplaceReferencesList {
+    private readonly list: ReplaceReferences[]
+
+    /** Reference name at the starting position. */
+    public readonly startRefName?: string | number
+
+    /** Reference name at the ending position. */
+    public readonly endRefName?: string | number
+
+    /** All reference names except at the starting position. */
+    public readonly otherThanStartRefNames: Set<string | number>
+
+    /** All reference names except at the starting position. */
+    public readonly otherThanEndRefNames: Set<string | number>
+
+    public constructor(list: ReplaceReferences[]) {
+        this.list = list
+        this.startRefName = list[0].startRef?.ref
+        this.endRefName = list[0].endRef?.ref
+
+        const otherThanStartRefNames = new Set<string | number>()
+        const otherThanEndRefNames = new Set<string | number>()
+        for (const { startRef, endRef, allRefs } of this.list) {
+            for (const ref of allRefs) {
+                if (ref !== startRef) {
+                    otherThanStartRefNames.add(ref.ref)
+                }
+                if (ref !== endRef) {
+                    otherThanEndRefNames.add(ref.ref)
+                }
+            }
+        }
+        this.otherThanStartRefNames = otherThanStartRefNames
+        this.otherThanEndRefNames = otherThanEndRefNames
+    }
+
+    public *[Symbol.iterator](): Iterator<ReplaceReferences> {
+        yield* this.list
+    }
+}
+
+const enum SideEffect {
+    startRef,
+    endRef,
+}
+
+/**
+ * Gets the type of side effect when replacing the capture group for the given element.
+ *
+ * If  the starting and ending capture groups, and the patterns between each have disjoints with each other,
+ * there are no side effects.
+ */
+function getSideEffectsWhenReplacingCapturingGroup(
+    start: CapturingGroup | undefined,
+    others: Element[],
+    end: CapturingGroup | undefined,
+    regexpContext: RegExpContext,
+): Set<SideEffect> {
+    const result = new Set<SideEffect>()
+    if (start && others.length) {
+        if (!hasDisjoint(start, ...others)) {
+            result.add(SideEffect.startRef)
+        }
```

Comment:
This is incorrect. Example: 
```js
> "aaaab".replace(/(a+)a+b/, "$1c")
'aaac'
> "aaaab".replace(/(?<=a+)a+b/, "c")
'ac'
```

We have to be really conservative when checking the elements after the capturing group. If we find an element that does not have a constant length (hint: [`getLengthRange`](https://rundevelopment.github.io/regexp-ast-analysis/docs/latest/modules.html#getlengthrange)), we have to stop (actually, we can get one additional character using `getFirstConsumedCharAfter`).

The function for determining disjointness should look something like this:

```js
const { char } = getPossiblyConsumedChar(start, flags)
for (const element of elements) {
  if (isConstantLength(element)) {
    if (char.isDisjointWith(getPossiblyConsumedChar(element, flags).char)) {
      return true;
    }
  } else {
    const look = FirstConsumedChars.toLook(getFirstConsumedCharPlusAfter(element, "ltr", flags));
    return char.isDisjointWith(look.char)
  }
}
return false
```

(See [`getFirstConsumedCharPlusAfter`](https://github.com/ota-meshi/eslint-plugin-regexp/blob/6e070304c6426b161cb61017ccd30945bdc7365f/lib/utils/regexp-ast/alternative-prefix.ts#L253))